### ISSUE TITLE
Stream bufferSize

### ISSUE THREAD

**Issue by casmeiron** [reactphp/react#324]
Hi, I'm using Ratchet to create a server socket that receives messages from a client.
The problem occurs when I try to send a message thats longer than 4096 bytes. The only way that I could make it work was changing (I don't like that) manually the bufferSize in Stream.php class. Is that a way to do that but calling some method acessors? Couldn't find out.

Just to give more insights, my server implements MessageComponentInterface and my onMessage its receiving only the first 4096 bytes of my message.

Thanks in advance.

**Comment by cboden** [comment]
`bufferSize` is a public variable on `Stream` you should just be able to change that. 

I don't think this is the problem you're having though. TCP is streamed and has its own chunking. 4096 is just the soft limit more than likely that's not even being reached. 

[This test case](http://socketo.me/reports/ab/ratchet_0_3_streams_case_9_3_9.html) shows Ratchet receiving and sending out WebSocket messages in chunks of 4MB. Internally, React is chunking that at 4096 bytes but it's being reassembled via the messaging protocol to the original 4MB frame.

**Comment by casmeiron** [comment]
I don't think I have access to Stream from where I stand.
here's the snippet of my code:
```php

$appServer = ...

$server = IoServer::factory($appServer, $port);

$appServer->setLoop($server->loop);

$server->run();

```php

And in my onMessage, I just got the message with only 4096bytes.
The $connection variable implements ConnectionInterface I don't have access to stream.

I see your test case but I also see my message arriving partially. Whats wrong?

Thanks in advance.

**Comment by drillcut** [comment]
Is connection drops on this check? Got same on post request with multipart/form-data content.
React/Stream/Buffer line 120:
if (0 === strlen($this->data)) {
            $this->loop->removeWriteStream($this->stream);
            $this->listening = false;

            $this->emit('full-drain', [$this]);
        }
If i comment this, server receive full content, if not it's chunked 8k (stored image size about 7k)

**Comment by Ethorsen** [comment]
I seem to have the same problem and can't find a solution to it. 

 I have two reactphp running and I'm using a Socket to have them communicate with each other.  I'm gzcompressing and base64_encoding any message transmitted between them.  The problem is that any message of more than 4096 are being cut in pieces and sent to the 'data' listener that way.

part of my code
```
	$connector->create('127.0.0.1', $socketPortNumber)->then(function (Stream $stream) use ($self, $tableId) {
		$self->outputInfo("Connected to table worker".$tableId, DEBUG);
		$stream->bufferSize = 32768;
		$stream->on('data', function ($data, $read) use ($self) {
			$self->handleTableWorkerOutput($data);
		});
	});
```

Even by setting the bufferSize to a higher value, I can't seem to get bigger data in one piece.  I could come up with a data split thingy and just reassemble on the other side, but I would just prefer to have this working as expected.

Any ideas on this?

**Comment by dancingshell** [comment]
@Ethorsen we were having the same problem and also tried to increase the bufferSize but it doesn't seem to make a difference. We solved it by concatenating the data packages before returning it to the calling function to handle. Not as clean as getting all the data in one package, but it gets the job done.
```
        $connector = new Connector($this->loop, $this->dns);
        $response = "";
        $connector->create($this->address, $this->port)->then(function ($stream) use (&$response) {
            $stream->on('data', function ($data) use (&$response) {
                if (!empty($data)) {
                    $response .= $data;
                }
            });

        });
        $this->loop->run();

        return $response;
```

**Comment by cboden** [comment]
Possibly fixed by reactphp/stream#20

Try with react/stream read_buffer branch.

**Comment by dancingshell** [comment]
@cboden Great that works, when I added that line in I was able to set the bufferSize higher and get the data in one package. Thanks.

**Comment by Ethorsen** [comment]
Thx @dancingshell and @cboden.  I did come up with my own split process where I would tag each part with a UID and a part # / total part.  Controller would wait to get all parts of a message before sending it for processing.

But anyhow I will try cboden fix later on.

**Comment by clue** [comment]
> * This test case shows Ratchet receiving and sending out WebSocket messages in chunks of 4MB. 
> * Possibly fixed by reactphp/stream#20

I feel this ticket has been answered and can be closed :+1: If we have new details, update this ticket and we can look into this again :+1:
